irohad/multi_sig_transactions/impl/gossip_propagation_strategy.cpp

```diff
@@ -0,0 +1,56 @@
+/**
+ * Copyright Soramitsu Co., Ltd. 2017 All Rights Reserved.
+ * http://soramitsu.co.jp
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "multi_sig_transactions/gossip_propagation_strategy.hpp"
+#include <numeric>
+#include <vector>
+
+namespace iroha {
+
+using PropagationData = PropagationStrategy::PropagationData;
+
+GossipPropagationStrategy::GossipPropagationStrategy(PropagationData data,
+                                                     uint32_t frequency,
+                                                     uint32_t amount)
+    : data(data),
+      emitent(
+          rxcpp::observable<>::interval(std::chrono::steady_clock::now(),
+                                        std::chrono::milliseconds(frequency))
+              .map([this, amount](int v) {
+                std::vector<decltype(data)::value_type> vec(amount);
+                std::for_each(vec.begin(), vec.end(), [this](auto &el) {
+                  if (this->non_visited.empty())
+                    this->initQueue();
+                  el = this->data[this->non_visited.top()];
+                  this->non_visited.pop();
+                });
+                return vec;
+              })) {}
```

Comment:
As I understand, each time we do not fetch random peers at any time, only in the first  tick. This is not required in the task and this is not reflected in a description. Pls, rework code or add a description with behavior.

Reply:
Random distribution may for some time skip some peers, so this distribution don't be uniform. This way shuffle the queue of peers when it empty and just pop elements from it
I'll add docs in the following commit

Reply:
Ok, also we need to reflect it in documentation further.